### PR TITLE
feat(cloudfront): add L2 support for CloudFront functions

### PR DESCRIPTION
**Pull request by hoegertn**
Add support for CloudFront functions

/cc @njlynch 

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-cloudfront/lib/distribution.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/distribution.ts:679]:
```suggestion
  /** The type of event which should invoke the function. */
```

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/function.ts:25]: Is there any reason to export this? If we change the signature of `fromInline` to just return `FunctionCode`, then this could be left un-exported, correct?

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/function.ts:102]: I believe these names need to be globally unique. Given that, we should probably have the name include the current region as well, so stacks with `Function`s can be deployed to multiple regions.

File: packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts:468]:
```suggestion
  /** The type of event which should invoke the function. */
```

File: packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts:984]: ❗ 

```suggestion
        functionAssociations: input.functionAssociations.map(association => ({
```

File: packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts
Comment (njlynch) [inline packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts:603]: Need this test for `Distribution` as well.

**Review by njlynch — CHANGES_REQUESTED** [review]
Thanks for this contribution!

It looks like the CloudFormation spec hasn't been updated yet, so this will need to wait until the next spec update. In the meantime, a few minor (and one important!) comments...

File: packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/lib/web-distribution.ts]: Copy & paste is dangerous ;-)

File: packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts:603]: There is an integ test for distribution. Is this ok or do you need the unit test too?

Comment (njlynch) [reply packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts:603]: A unit test too, please. Unit tests are run much more frequently and are easier to maintain. In my perspective, at least, everything should have a unit test to verify intended functionality, and each resource (or resource-link) an integration test to verify deployability.

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (jogold) [inline packages/@aws-cdk/aws-cloudfront/lib/function.ts:8]: Would add a `public static fromFile(path: string)` here

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (jogold) [inline packages/@aws-cdk/aws-cloudfront/lib/function.ts:64]: looks from the specs that `comment` is required after all?

(not in the console but in the CF docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cloudfront-function-functionconfig.html#cfn-cloudfront-function-functionconfig-comment)

**Comment by njlynch** [comment]
Hi @hoegertn ; just wanted to note I'm out for the next week. I think if we can get this working with just the `FunctionARN` as an attribute, and add the `Stage` later, I think that'd be fine. If you're unblocked and want to push this through, please feel free to tag one of my colleagues to look at this in my absence.

**Comment by hoegertn** [comment]
Ok, according to the team, CFN support should be fixed this week so I hope I can finish the PR this weekend.

**Comment by cornelcroi** [comment]
Hi.
Any idea when this PR will be merged ?

**Comment by hoegertn** [comment]
> Hi.
> Any idea when this PR will be merged ?
> 
> Thanks.

As I mentioned above the CFN support currently is broken. As soon as this is fixed by the CloudFront team I can finalize this PR and get it merged.

**Comment by robertd** [comment]
> Ok, according to the team, CFN support should be fixed this week so I hope I can finish the PR this weekend.

@hoegertn Is this stated somewhere in the cfn-roadmap repo?

**Comment by hoegertn** [comment]
No, I am in close contact with the CloudFront team. This is implemented by the service teams and not the CFN team.

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/lib/function.ts:143]: Done

File: packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/test/web-distribution.test.ts:603]: Done

**Comment by florianwiech** [comment]
@njlynch @NetaNir @iliapolo Is something missing or blocking the PR from being merged? I am currently waiting for it.

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-cloudfront/lib/function.ts:28]: Still think we should default to leaving this un-exported until/unless we have a good reason to export it.

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-cloudfront/lib/function.ts:8]: Would definitely be a nice-to-have; not a blocker for approval, but certainly wouldn't hurt.

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (njlynch) [reply packages/@aws-cdk/aws-cloudfront/lib/function.ts:83]: Given that, let's update the docs.

```suggestion
   * @default - same as `functionName`
```

**Review by njlynch — DISMISSED** [review]
Sorry for the delay. Back at it again! :)

Looks good. One minor doc update and the note about `InlineCode`, and then we can get this merged!

File: packages/@aws-cdk/aws-cloudfront/lib/function.ts
Comment (hoegertn) [reply packages/@aws-cdk/aws-cloudfront/lib/function.ts:8]: I will add this later. I want to do this using an extra class for it and want to think about the API. I would love to have esbuild/bundling here but given that everything is ES5 this is not trivial.

**Review by njlynch — DISMISSED** [review]
Hooray! Thanks for all the effort to get this done. 🎉

**Review by njlynch — APPROVED** [review]

**Pull request merged**
